### ISSUE TITLE
Training with FP8

### ISSUE THREAD

**Issue by casper-hansen** [axolotl-ai-cloud/axolotl#755]
### ⚠️ Please check that this feature request hasn't been suggested before.

- [X] I searched previous [Ideas in Discussions](https://github.com/OpenAccess-AI-Collective/axolotl/discussions/categories/ideas) didn't find any similar feature requests.
- [X] I searched previous [Issues](https://github.com/OpenAccess-AI-Collective/axolotl/labels/enhancement) didn't find any similar feature requests.

### 🔖 Feature description

The idea is to replace the Linear and Norm layers with the equivalent FP8 layer from [Transformer Engine](https://github.com/NVIDIA/TransformerEngine). This should give us a 20% boost in training speed when doing full finetuning.

FP8 is supported by the Ada Lovelace architecture and later (SM89+). In practice, that means we have RTX 4090, L40, and H100 GPUs that can use FP8.

Loss is supposed to be identical to FP16:

![image](https://github.com/OpenAccess-AI-Collective/axolotl/assets/27340033/c7ada5e0-ed6a-4df6-9b8c-12cb4b6556ad)

### ✔️ Solution

Get inspiration from LLM Foundry:

https://github.com/mosaicml/llm-foundry/pull/271
https://github.com/mosaicml/llm-foundry/pull/432

Accelerate implements custom padding for use with Transformer Engine and FP8:
https://github.com/huggingface/accelerate/blob/c7c99a30ea88f00c854017d362e72b9cf60c7db6/examples/nlp_example.py#L80

`mixed_precision="fp8"` to the accelerator:
https://github.com/huggingface/accelerate/blob/c7c99a30ea88f00c854017d362e72b9cf60c7db6/examples/complete_nlp_example.py#L55

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this feature has not been requested yet.
- [X] I have provided enough information for the maintainers to understand and evaluate this request.

**Comment by dakshvar22** [comment]
Hi, any update on when fp8 aware fine-tuning would be available?

**Comment by ehartford** [comment]
@winglian this is important for training DeepSeek-V3 and DeepSeek-R1

```
"quantization_config": {
  "activation_scheme": "dynamic",
  "fmt": "e4m3",
  "quant_method": "fp8",
  "weight_block_size": [
    128,
    128
  ]
},
```

**Comment by ehartford** [comment]
Maybe this is relevant?

https://huggingface.co/docs/accelerate/usage_guides/low_precision_training

**Comment by ehartford** [comment]
This too

https://github.com/huggingface/accelerate/tree/main/benchmarks/fp8

**Comment by ehartford** [comment]
the DeepSpeed example, 

with ms_amp:
https://github.com/huggingface/accelerate/blob/main/benchmarks/fp8/ms_amp/distrib_deepspeed.py

with transformer_engine:
https://github.com/huggingface/accelerate/blob/main/benchmarks/fp8/transformer_engine/distrib_deepspeed.py

**Comment by mhenrichsen** [comment]
Marking as stale.
